internal/cephfs/core/snapshot_metadata.go

```diff
@@ -0,0 +1,77 @@
+package core
+
+import (
+	"fmt"
+	"strings"
+)
+
+// GetSnapshotMetadata gets custom metadata on the subvolume snapshot in a
+// volume based on key name.
+func (s *snapshotClient) GetSnapshotMetadata(key string) (string, error) {
+	fsa, err := s.conn.GetFSAdmin()
+	if err != nil {
+		return "", err
+	}
+
+	return fsa.GetSnapshotMetadata(s.FsName, s.SubvolumeGroup, s.VolID, s.SnapshotID, key)
+}
+
+// SetSnapshotMetadata sets custom metadata on the subvolume snaphot in a
+// volume as a key-value pair.
+func (s *snapshotClient) SetSnapshotMetadata(key, value string) error {
+	fsa, err := s.conn.GetFSAdmin()
+	if err != nil {
+		return err
+	}
+
+	return fsa.SetSnapshotMetadata(s.FsName, s.SubvolumeGroup, s.VolID, s.SnapshotID, key, value)
+}
+
+// RemoveSnapshotMetadata removes custom metadata set on the subvolume
+// snapshot in a volume using the metadata key.
+func (s *snapshotClient) RemoveSnapshotMetadata(key string) error {
+	fsa, err := s.conn.GetFSAdmin()
+	if err != nil {
+		return err
+	}
+
+	return fsa.RemoveSnapshotMetadata(s.FsName, s.SubvolumeGroup, s.VolID, s.SnapshotID, key)
+}
+
+// ListSnapshotMetadata lists custom metadata (key-value pairs) set on the subvolume
+// snapshot in a volume.
+func (s *snapshotClient) ListSnapshotMetadata() (map[string]string, error) {
+	fsa, err := s.conn.GetFSAdmin()
+	if err != nil {
+		return nil, err
+	}
+
+	return fsa.ListSnapshotMetadata(s.FsName, s.SubvolumeGroup, s.VolID, s.SnapshotID)
+}
+
+// SetAllSnapshotMetadata set all the metadata from arg parameters on
+// Subvolume snapshot.
+func (s *snapshotClient) SetAllSnapshotMetadata(parameters map[string]string) error {
+	for k, v := range parameters {
+		err := s.SetSnapshotMetadata(k, v)
+		if err != nil {
+			return fmt.Errorf("failed to set metadata key %q, value %q on subvolume snapshot %v: %w", k, v, s, err)
+		}
+	}
+
+	return nil
+}
+
+// UnsetAllSnapshotMetadata unset all the metadata from arg keys on subvolume
+// snapshot.
+func (s *snapshotClient) UnsetAllSnapshotMetadata(keys []string) error {
+	for _, key := range keys {
+		err := s.RemoveSnapshotMetadata(key)
+		// TODO: replace string comparison with errno.
```

Comment:
time to address `TODO` we have this too for all the metadata operations

Reply:
This must be coming from Ceph, Ceph commands are initiated from go-ceph to Ceph and go-ceph actually unmarshal the json output passed from Ceph. So where should be searching for a matching error? That is something I don't know :-(
I'm happy for the adventure but need some lead here.

Maybe @phlogistonjohn @nixpanic might know.